### PR TITLE
[hop] Support non-node outputs (constants) in subgraph output

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #162702
* __->__ #162610
* #161458

eg when tracing this graph:
```python
def forward(self, boosted_weight : torch._subclasses.fake_tensor.FakeTensor, to : torch._subclasses.fake_tensor.FakeTensor):
    t = boosted_weight.t();  boosted_weight = None
    matmul = torch.matmul(to, t);  to = t = None
    return (matmul,)
```

when partitioning its joint, and arg1_1 is a fw_input that doesn't require gradients:
```python
(Pdb++) joint_hop_gm.print_readable()
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "bf16[6144, 6144]", arg1_1: "bf16[64, 256, 6144]", arg2_1: "bf16[64, 256, 6144]"):
        # No stacktrace found for following nodes
        t: "bf16[6144, 6144]" = torch.ops.aten.t.default(arg0_1);  arg0_1 = None
        view: "bf16[16384, 6144]" = torch.ops.aten.view.default(arg1_1, [16384, 6144]);  arg1_1 = None
        mm: "bf16[16384, 6144]" = torch.ops.aten.mm.default(view, t);  t = None
        _unsafe_view: "bf16[64, 256, 6144]" = torch.ops.aten._unsafe_view.default(mm, [64, 256, 6144]);  mm = None
        view_1: "bf16[16384, 6144]" = torch.ops.aten.view.default(arg2_1, [16384, 6144]);  arg2_1 = None
        t_1: "bf16[6144, 16384]" = torch.ops.aten.t.default(view_1);  view_1 = None
        mm_1: "bf16[6144, 6144]" = torch.ops.aten.mm.default(t_1, view);  t_1 = view = None
        t_2: "bf16[6144, 6144]" = torch.ops.aten.t.default(mm_1);  mm_1 = None
        t_3: "bf16[6144, 6144]" = torch.ops.aten.t.default(t_2);  t_2 = None
        return (t_3, None, _unsafe_view)
```

### CODE REVIEW

**Comment by bdhirsh** [comment]
what's the use case for returning constants? I would have imagined that in any scenario where we have a subgraph returning a constant, we could just have inlined that constant anywhere it's needed in the outer graph (and not have to return it in the subgraph)

**Comment by xmfan** [comment]
This is for the backward subgraph of the HOP, where it sometimes returns None, I'm guessing for an input that does not require gradients
